### ISSUE TITLE
.NET May 2021 Update - .NET 5.0.6, .NET Core 3.1.15 and .NET Core 2.1.28

### ISSUE THREAD

**Issue by rbhanda** [dotnet/core#6247]
Release Notes 
[5.0.6](https://github.com/dotnet/core/blob/master/release-notes/5.0/5.0.6/5.0.6.md)
[3.1.15](https://github.com/dotnet/core/blob/master/release-notes/3.1/3.1.15/3.1.15.md)
[2.1.28](https://github.com/dotnet/core/blob/master/release-notes/2.1/2.1.28/2.1.28.md)

Please report any issues you find either by responding to this issue, creating a new issue or creating a new issue in one of the following repos:

* [dotnet/aspnetcore](https://github.com/dotnet/aspnetcore/issues) - for ASP.NET Core questions
* [dotnet/efcore](https://github.com/dotnet/efcore/issues) - for Entity Framework Core questions
* [dotnet/sdk](https://github.com/dotnet/sdk/issues) - for CLI tools and questions
* [dotnet/runtime](https://github.com/dotnet/runtime/issues) - for runtime, API, or installer issues and questions
* [dotnet/winforms](https://github.com/dotnet/winforms/issues) - for WinForms issues
* [dotnet/wpf](https://github.com/dotnet/wpf/issues) - for WPF issues
* [nuget/home](https://github.com/nuget/home/issues) - for NuGet questions and issues

-----
Status of availability of SDK on Microsoft's Linux feeds: 

| Distro | 5.0.6 | 3.1.15 | 2.1.28 | 
|--|--|--|--|
| Ubuntu 16.04   |  ✅  |  ✅  |  ✅  | 
| Ubuntu 18.04    |  ✅  |  ✅  |  ✅  | 
| Ubuntu 20.04    |  ✅  |  ✅  |  ✅  | 
| Ubuntu 20.10    |  ✅  |  ✅  |  ✅  | 
| Ubuntu 21.04     |  ✅  |  ✅  |  ✅  | 
| Centos 7      |  ✅  |  ✅  |  ✅  | 
| Debian 9   |  ✅  |  ✅  |  ✅  | 
| Debian 10   |  ✅  |  ✅  |  ✅  | 
| Fedora 32    |  ✅  |  ✅  |  ✅  | 
| Fedora 33    |  ✅  |  ✅  |  ✅  | 
| Fedora 34  |  ✅  |  ✅  |  ✅  | 
| OpenSUSE 15   |  ✅  |  ✅  |  ✅  | 
| SLES 12   |  ✅  |  ✅  |  ✅  | 
| SLES15  |  ✅  |  ✅  |  ✅  | 

Note: This list refers to the Microsoft-provisioned feeds (packages.microsoft.com) and does not in any way represent direct availability in distros (eg RHEL, Fedora).

### Known Issues
If there are any issues with this release we will track them here and check issues off as they're resolved. See the linked issues for details on progress and resolution details.

- [x]  Incorrect VS version displayed for .NET Core 2.1.816 https://github.com/dotnet/core/issues/6247#issuecomment-839231100
- [x] dotnet-sdk-3.1.115-x64 rpm missing from https://packages.microsoft.com/rhel/7/prod/

**Comment by KalleOlaviNiemitalo** [comment]
The download page <https://dotnet.microsoft.com/download/dotnet/2.1> says:

> ### SDK 2.1.816
>
> **Visual Studio support**\
> Visual Studio 2017 (v15.9)

Should that be Visual Studio 2019 (v16.2) instead, like in SDK 2.1.815?

**Comment by leecow** [comment]
Yes, this should be 16.2. Looks like it's correct in https://github.com/dotnet/core/blob/main/release-notes/2.1/releases.json. @mairaw can you have a look?

**Comment by rbhanda** [comment]
I found the problem. https://github.com/dotnet/core/pull/6252 to fix the VS version issue

**Comment by KalleOlaviNiemitalo** [comment]
TLDR: I got NU1403 errors after the May 2021 updates, but clearing `%USERPROFILE%\.nuget\packages` fixed them.

After installing:

- .NET SDK 5.0.203 (x64)
- .NET Core SDK 2.1.524 (x64)
- Visual Studio Build Tools 2017 version 15.9.36, including .NET Core build tools

I got errors like the following when building a C# project:

- error NU1403: Package content hash validation failed for Microsoft.NETCore.Targets.1.1.0. The package is different than the last restore.
- error NU1403: Package content hash validation failed for Microsoft.Win32.Registry.4.3.0. The package is different than the last restore.
- error NU1403: Package content hash validation failed for runtime.native.System.4.3.0. The package is different than the last restore.
- error NU1403: Package content hash validation failed for System.Runtime.Serialization.Primitives.4.1.1. The package is different than the last restore.

This C# project has `<Project Sdk="Microsoft.NET.Sdk">` and `<TargetFramework>netstandard2.0</TargetFramework>`. It is built with `dotnet build`. There is no global.json file, so the build used the latest .NET SDK 5.0.203. There are no editions of Visual Studio 2019 on the computer.

Process Monitor (filtered with "nuget") showed dotnet.exe accessing:
- `%USERPROFILE%\.nuget\packages\microsoft.netcore.targets\1.1.0\microsoft.netcore.targets.1.1.0.nupkg.sha512`, which contained:
  ```
  Bev/m4CGisEJzE5/OP9onHygFaDgTJ4Q/kxhvdvaSDzNEIkaUi2T3GPMgOgh9MvSrjH3JTcZMNxuqzDVcfZmeQ==
  ```
- `C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.targets\1.1.0\.nupkg.metadata`, which contained:
  ```json
  {
    "version": 1,
    "contentHash": "aOZA3BWfz9RXjpzt0sRJJMjAscAUm3Hoa4UWAfceV9UTYxgwZ1lZt5nO2myFf+/jetYQo4uTP7zS8sJY67BBxg=="
  }
  ```
- `%USERPROFILE%\.nuget\packages\microsoft.netcore.targets\1.1.0\.nupkg.metadata`, which contained:
  ```json
  {
    "version": 2,
    "contentHash": "KxBX5BxEXSm/2rijX7KL4Sg3wA2XpgQ4y12Mup3NIlZO3YBsWkA149fZIBb9LjMo8ehABFuWkAwyLxtu+MRT0g==",
    "source": "C:\\Program Files\\dotnet\\sdk\\NuGetFallbackFolder"
  }
  ```

Other related files:

- `packages.lock.json` in the project included:
  ```json
        "Microsoft.NETCore.Targets": {
          "type": "Transitive",
          "resolved": "1.1.0",
          "contentHash": "aOZA3BWfz9RXjpzt0sRJJMjAscAUm3Hoa4UWAfceV9UTYxgwZ1lZt5nO2myFf+/jetYQo4uTP7zS8sJY67BBxg=="
        },
  ```
- '%USERPROFILE%\.nuget\packages\microsoft.netcore.targets\1.1.0\microsoft.netcore.targets.1.1.0.nupkg.sha512' contained:
  ```
  Bev/m4CGisEJzE5/OP9onHygFaDgTJ4Q/kxhvdvaSDzNEIkaUi2T3GPMgOgh9MvSrjH3JTcZMNxuqzDVcfZmeQ==
  ```
- The SHA-512 of `%USERPROFILE%\.nuget\packages\microsoft.netcore.targets\1.1.0\microsoft.netcore.targets.1.1.0.nupkg` was:
  ```
  05ebff9b80868ac109cc4e7f38ff689c7ca015a0e04c9e10fe4c61bddbda483ccd10891a522d93dc63cc80e821f4cbd2ae31f725371930dc6eab30d571f66679
  ```
  which corresponds to Base64:
  ```
  Bev/m4CGisEJzE5/OP9onHygFaDgTJ4Q/kxhvdvaSDzNEIkaUi2T3GPMgOgh9MvSrjH3JTcZMNxuqzDVcfZmeQ==
  ```
  i.e. it matched the `microsoft.netcore.targets.1.1.0.nupkg.sha512` file in the same directory.
- `C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.targets\1.1.0\microsoft.netcore.targets.1.1.0.nupkg.sha512` contained:
  ```
  HvAzpoaIqrmZfsHAN4rLFji0r7YY5TP8r3SdkziXN7qU9KCpRIG+zfcBx+mIri/jkBNqjq4iWIfuYNtFBjSQ/g==
  ```
- The SHA-512 of `C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.targets\1.1.0\microsoft.netcore.targets.1.1.0.nupkg` was:
  ```
  8ba33b019f7a1ad8c979315a2b066383a952a26bede768b43b2c9a7331242b9ac62e6d6b5d9e6be439cac138270893ede502bb40278c8865ee076ea5556b4437
  ```
  which corresponds to Base64:
  ```
  i6M7AZ96GtjJeTFaKwZjg6lSomvt52i0OyyaczEkK5rGLm1rXZ5r5DnKwTgnCJPt5QK7QCeMiGXuB26lVWtENw==
  ```
  i.e. it did not match the `microsoft.netcore.targets.1.1.0.nupkg.sha512` file in the same directory.

I then ran repairs on .NET Core SDK 2.1.524 (x64) and .NET SDK 5.0.203 (x64), in that order. This did not change `C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.targets\1.1.0\microsoft.netcore.targets.1.1.0.nupkg.sha512` nor `C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.targets\1.1.0\microsoft.netcore.targets.1.1.0.nupkg`.

I then renamed `%USERPROFILE%\.nuget\packages` to `%USERPROFILE%\.nuget\packages~` and tried building the C# project again. This time, it succeeded. I had not changed `packages.lock.json`. After the build, there was no `%USERPROFILE%\.nuget\packages\microsoft.netcore.targets` directory.

I'm not blocked by this now, but I'd like to understand:

- Which installer manages this `C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.targets\1.1.0` directory?
- Is it a bug that `C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.targets\1.1.0\microsoft.netcore.targets.1.1.0.nupkg.sha512` does not match `C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.targets\1.1.0\microsoft.netcore.targets.1.1.0.nupkg`?
- Should I proactively delete the per-user NuGet package cache after each update of .NET SDK or Visual Studio Build Tools?
- Should I have posted this in [nuget/home](https://github.com/nuget/home/issues) instead?

**Comment by KalleOlaviNiemitalo** [comment]
I tried `nuget.exe verify -signatures` with NuGet 5.9.1.11, on `C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.targets\1.1.0\microsoft.netcore.targets.1.1.0.nupkg` at the build server, and got the same error as in <https://github.com/NuGet/Home/issues/7414>:

```
NU3005: The package signature file entry is invalid. The central directory header field 'compression method' has an invalid value (8).
```

So, I think the sha512 mismatch is a red herring.

**Comment by securylight** [comment]
I need to remove hosting bundle and no matter what I tried I can't remove it.

>dotnet-core-uninstall remove 3.1.15 --hosting-bundle
The following items will be removed:
  Microsoft .NET Core 3.1.15 - Windows Server Hosting

To avoid breaking Visual Studio or other problems, read https://aka.ms/dotnet-core-uninstall-docs.

Do you want to continue? [y/n] y
Uninstalling: Microsoft .NET Core 3.1.15 - Windows Server Hosting.
The uninstallation operation failed. Exit code was -1073741819 for ""C:\ProgramData\Package Cache\{0453c7eb-3633-4d61-80c1-6117bc512160}\WindowsServerHostingBundle.exe" /uninstall /quiet".

**Comment by rbhanda** [comment]
closing in favor of https://github.com/dotnet/core/issues/6337
